### PR TITLE
Add toXContent and fromXContent for DiscoveryNode and DiscoveryNodes

### LINKED ISSUE #13691: [Remote Cluster State] Add toXContent and fromXContent for DiscoveryNodes
### Is your feature request related to a problem? Please describe

Parent project: Cluster State Publication using Remote State
Meta issue: #13683

### Describe the solution you'd like

To upload and recreate the cluster state after downloading the state from remote, we need to have proper serialization and de-serialization methods for the DiscoveryNodes object. So, need to update the current toXContent methods for DiscoveryNodes and DiscoveryNode classes to include all necessary attribute which are needed to recreate the object when we need to recreate the object from remote.

### Related component

Cluster Manager

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shiv0408**
### Description
We need to upload the DiscoveryNodes object to remote to enable the cluster state publication flow from remote. To do that we need to parse DiscoveryNodes object to and from XContent.

### Related Issues
Resolves #13691 
Meta issue #13683 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by shiv0408** [comment]
`SmokeTestMultiNodeClientYamlTestSuiteIT` test has been identified as flaky previously in #9332

File: server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java [hunk 1]
```diff
@@ -544,21 +556,97 @@ public String toString() {
...
+        Set<DiscoveryNodeRole> roles = new HashSet<>();
+        Version version = null;
+        String currentFieldName = parser.currentName();
+        // token should be start object at this point
+        // XContentParser.Token token = parser.nextToken();
+        // if (token != XContentParser.Token.START_OBJECT) {
+        // throw new IllegalArgumentException("expected object but got a " + token);
+        // }
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:610]: This doesn't seem to align with your implementation of `ToXContentFragment` (which doesn't require start/end object) as well as the code on lines 595-597 which implies the `START_OBJECT` is optional.  You'd hit the end of the parser without getting the `END_OBJECT`. (Same comment on `DiscoveryNodes` class below.)

File: server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java [hunk 2]
```diff
@@ -544,21 +556,97 @@ public String toString() {
...
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:612]: This parsing isn't very readable. You have multi-branch if/else wrapped inside other multi-branch if/else.  I'm sure somewhere there's a complexity metric that would scream at you. :)

You always will have a field name (or it's invalid XContent) so if you're expecting the field name you should just be able to get the name. 

After that you can use a switch statement on the field names since you know their data types as well (or it's invalid and will throw an exception if it's the wrong type).

Something like
```java
while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
    String fieldName = parser.currentName();
    parser.nextToken();
    switch (fieldName) {
        case KEY_NAME:
            nodeName = parser.text();
            break;
        case KEY_EPHEMERAL_ID:
            ephemeralId = parser.text();
            break;
        // and so on.
```

Same comment on `DiscoveryNodes` class below.

File: server/src/test/java/org/opensearch/cluster/node/DiscoveryNodeTests.java [hunk 3]
```diff
@@ -249,4 +257,70 @@ public void testDiscoveryNodeIsSearchNode() {
...
+            Settings.EMPTY,
+            new TransportAddress(TransportAddress.META_ADDRESS, 9200),
+            "node_1"
+        );
+        XContentBuilder builder = JsonXContent.contentBuilder().prettyPrint();
+        builder.startObject();
+        node.toXContent(builder, new ToXContent.MapParams(singletonMap(Metadata.CONTEXT_MODE_PARAM, CONTEXT_MODE_GATEWAY)));
+        builder.endObject();
+
+        String expectedNodeAPUXContent = "{\n"
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/cluster/node/DiscoveryNodeTests.java:299]: If you really want to test against ToXContentFragment interface you need to test partial JSON that doesn't start and end with `{` and `}`.  (You might find that your tests currently fail!)  

Or require start and end and implement `ToXContentObject` and this test is OK.

File: server/src/test/java/org/opensearch/cluster/node/DiscoveryNodesTests.java [hunk 4]
```diff
@@ -497,6 +511,164 @@ public void testMaxMinNodeVersion() {
...
+                    expectedNodeAPUXContent,
+                    "    ",
+                    id,
+                    entry.getValue().getEphemeralId(),
+                    entry.getValue().getAddress().toString(),
+                    entry.getValue().getAttributes().isEmpty()
+                        ? " "
+                        : "\n" + "        \"custom\" : \"" + entry.getValue().getAttributes().get("custom") + "\"\n      "
+                );
+            }).collect(Collectors.joining(",\n")) + "\n" + "  }\n" + "}"
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/cluster/node/DiscoveryNodesTests.java:542]: Holy readability, Batman!  Switching from concatenation to a stream with some concatenation and `String.format()` combined....  I know it's a test class but even then I'm not sure what it's testing without a comment explaining what this is supposed to look like. This code here might be screaming for a `StringBuilder` implementation. Or a comment like you have below in `getExpectedXContentInGatewayMode`.

**Review by dbwiddis — COMMENTED** [review]
A few comments. 

Also I've seen the method you call `fromXContent` called `parse()`.  Since it's a static method and not enforced by the interface, in theory you can make it whatever you want... I see a lot of both throughout the codebase so either is probably fine, but I'd suggest making this consistent with other code around where it will be used... if that's `fromXContent()` that's fine, I haven't dug into the context.

File: server/src/test/java/org/opensearch/cluster/node/DiscoveryNodesTests.java (hunk 4 above)
Comment (shiv0408) [reply server/src/test/java/org/opensearch/cluster/node/DiscoveryNodesTests.java:542]: I know this was bad when I wrote this. To keep the random discovery node, it was really difficult to create a string which was readabile enough. Let me give StringBuilder a try.

File: server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java (hunk 2 above)
Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:612]: I agree @dbwiddis, I was bothered by these nested if/else, but that's how we were doing this in many other classes, so i didn't change the approach. But, will update this to use switch statement like you suggested.

File: server/src/test/java/org/opensearch/cluster/node/DiscoveryNodesTests.java (hunk 4 above)
Comment (dbwiddis) [reply server/src/test/java/org/opensearch/cluster/node/DiscoveryNodesTests.java:542]: It's ok as is if you just give an example of what it builds.

File: server/src/test/java/org/opensearch/cluster/node/DiscoveryNodeTests.java (hunk 3 above)
Comment (shiv0408) [reply server/src/test/java/org/opensearch/cluster/node/DiscoveryNodeTests.java:299]: Actually the start and end `{` `}` are coming from line 295, 297. To build a valid json, we need to have start and end objects in our builder or it will fail to create a Json string. As the start and end of json are coming from line 295 and 297, our toXContent is only returning the partial XContent fragment which is getting verified in our test.

**Comment by lukas-vlcek** [comment]
@shiv0408 Would you mind modifying issue description and adding a ticket number that requires this change please? I would be interested in learning more about why exactly this is needed (I assume this is relevant to remote storage?).

File: server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java (hunk 1 above)
Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:610]: I believe there might be some misunderstanding here. I agree that DiscoveryNode object's XContent does not need to have a start and end object, but here we are checking for the end object which is added in line 581 above

Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:610]: But for `DiscoveryNodes`, this comment does make sense. I will modify the class to implement ToXContentObject.

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:610]: > I believe there might be some misunderstanding here. I agree that DiscoveryNode object's XContent does not need to have a start and end object

I'm not sure this is the desired state here. Why don't you want to extend `ToXContentObject` instead of `ToXContentFragment`?

> but here we are checking for the end object which is added in line 581 above

You are also adding a start object.  Which is fine as long as they're paired.  But you can't guarantee (based on the interface type) that you're always receiving the same object you wrote.  It's possible someone else may make a larger object and include this inline -- the purpose of `ToXContentFragment`.

You make the start object optional here:
```java
        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
            parser.nextToken();
        }
```

So if someone else trusts your object inheritance and places the object inline without start and end objects then you'd never get to the end and would keep iterating `parser.nextToken()`.  Eventually that will be null, and null won't equal end object, so inside the loop you'll throw an NPE on `token.isValue()`.

So you need to either test for a null `nextToken()` in addition to the END_OBJECT as a loop stopping condition, or you need to implement/extend `ToXContentObject` which requires start and end objects.   I suspect you probably want the latter.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:606]: Actually, `DiscoveryNode` is part of the `DiscoveryNodes` toXContent response and getting used as a fragment like you mentioned and I plan to use `DiscoveryNodes` toXContent in ClusterState going forward. That's the reason for keeping them Fragment. Anyway, I have pushed a commit modifying the toXContent method. If you still believe we can hit a NPE in the current method. Can you please provide a sample json, I would write a unit test based on that to ensure that NPE is never thrown from our function.

**Comment by shiv0408** [comment]
@lukas-vlcek apologies for late response, I have modified the description of the PR to include the issue for this. And yes, this effort is required for remote state publication.

**Comment by shiv0408** [comment]
@dbwiddis Thanks for reviewing the changes. I have pushed a commit to address your concerns. Can you please take another look at the changes? 

> Also I've seen the method you call fromXContent called parse()
We are consistently using `fromXContent` method name for de-serializing the object for all the objects with are getting used in our remote cluster state flow.

File: server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java (hunk 1 above)
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:605]: > Actually, `DiscoveryNode` is part of the `DiscoveryNodes` toXContent response and getting used as a fragment

No, it's being used as an object.  Your example in a comment below shows how it's used inline, and it includes th e `{` (START_OBJECT) and `}` (END_OBJECT).  It's a nested complete object, not a fragment.
```json
"node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
}
```
For a fragment example, see the `TransportAddress` class.  Its `toXContent` method simply adds the string value, without the object wrapper:
```java
@Override
public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
    return builder.value(toString());
}   
```
This is a trivial example, writing the `toString()` (which is what you read when parsing) but is more commonly used for creating arrays.... you'd have START_ARRAY and several objects and then END_ARRAY.  [Here is an example](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/transport/DeleteModelResponse.java).

In your builder, you always start and end with the start object/end object, so as long as you're always parsing JSON that you created with your own `toXContent()` method, your code will work.   

So I have to be a bit creative to get malformed JSON but here's an example: The problem is that by making start object optional, you would allow malformed JSON like this to partially parse:
```json
{ [
  "node_1" : 
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
] }
```
Note the start object external to the array at the start. You read the field name and then try to read this object.  Since you've made start object optional, it jumps right in and parses name, ephemeral_id, transport_address, and attributes.   it's still looking for the `}` but doesn't find it, instead it finds the end array. With your current code it would throw an exception on an unexpected token.

If you didn't have the array and did have an enclosing outer object, you'd now be mismatched as you'd have read the closing object for something else other than your own object, and would get some other confusing parsing error that didn't really match the _real_ error, the omission of the `{` at the start of this object.

TLDR:
1. you _must_ match a required end object to a required start object.  You require one but make the other one optional.
2. you are implementing `ToXContentFragment` but you really should implement `ToXContentObject` here.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:605]: > you are implementing ToXContentFragment but you really should implement ToXContentObject here.

https://github.com/opensearch-project/OpenSearch/blob/9b9158e4cb692dfe04315225ce2b5f9c6c453307/server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java#L69

I am not changing the implementing `ToXContentFragment` interface in the `DiscoveryNode`. It was already an implementation of the this interface and changing it to a `ToXContentObject` might causing breaking changes in our multiple API responses. As this object is already getting used as a inline object in cluster state API, and other cat APIs.

>No, it's being used as an object. Your example in a comment below shows how it's used inline, and it includes th e { (START_OBJECT) and } (END_OBJECT). It's a nested complete object, not a fragment.

Even a nested object is a fragment only as there is no start and end object around it. According to our documentation [here](https://github.com/opensearch-project/OpenSearch/blob/9b9158e4cb692dfe04315225ce2b5f9c6c453307/libs/core/src/main/java/org/opensearch/core/xcontent/ToXContentFragment.java#L36-L41), a fragment needs to have start and end object externally.
The example of you gave around fragment being used for creating array. My use case is also similar, where instead of create a field which has a `START_ARRAY`, our object has `START_OBJECT`.

The JSON you have provided is a invalid. And we don't expect to parse XContent not created from our method. Even if our object is getting used inline, like I am using in `DiscoveryNodes` class, the XContent is created from our toXContent method and object is parsed from our fromXContent method. 

> you must match a required end object to a required start object. You require one but make the other one optional

I am matching start and end object which are being added in our toXContent method. The line 590-592 you are referring, is to enable parsing of following object. This is a complete XContent object, but even if we don't receive the complete object (i.e. without start and end), we will be able to parse that inline object.
```
{
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
}
```

File: server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java [hunk 5]
```diff
@@ -544,21 +557,107 @@ public String toString() {
...
+            if (token.isValue()) {
+                switch (currentFieldName) {
+                    case KEY_NAME:
+                        nodeName = parser.text();
+                        break;
+                    case KEY_EPHEMERAL_ID:
+                        ephemeralId = parser.text();
+                        break;
+                    case KEY_TRANSPORT_ADDRESS:
+                        transportAddress = TransportAddress.fromString(parser.text());
```
[the remark this answers is not in the document]
Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:618]: Don't think we will be able to do that now. I will modify this and add a test to ensure that TransportAddress.fromString can parse both ipv4 and ipv6 addresses.

File: server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java (hunk 1 above)
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:605]: > I am not changing the implementing `ToXContentFragment` interface in the `DiscoveryNode`. It was already an implementation of the this interface and changing it to a `ToXContentObject` might causing breaking changes 

Agreed, you shouldn't change this.

> Even a nested object is a fragment only as there is no start and end object around it. 

Agreed.... but then you check for the end object in the line prompting this comment chain:
```java
 while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) 
```
If there is no end object, this loop will never reach its stopping condition.

> The JSON you have provided is a invalid.

I know. But the parsing error thrown would have been somewhere unrelated to the invalid JSON, in some follow-on code much more confusing to debug.  I just threw up a quick example to show that in the case you do not require a start object, and thus shouldn't require an end object (because they should match) you end up with the code as it was written hitting some later end object.  Or possibly getting a NPE if you're just parsing a small fragment with no start or end.


> > you must match a required end object to a required start object. You require one but make the other one optional
> 
> I am matching start and end object which are being added in our toXContent method. 

When you create the JSON you can match them. If you are reading the JSON created elsewhere you do not control that.  Someone else could be creating that JSON outside of core.

The code around the loop looks better because you're matching the start (lines 603-604) to the end (the stopping condition for the loop in 605).

But there is still an optional `START_OBJECT` present in lines 590-592 without a corresponding end object.  I still think this is wrong.

It should be legal for anyone parsing this JSON to parse their own start/end outside of your parsing.  SO taking your JSON above:
```json
{
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
}
```
it should be legal to pass this (valid JSON) string as a fragment which completely encapsulates all the information in the object, lacking only its "field name" which is only relevant for `DiscoveryNodes`.
```json
  {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
```
This fails, however, because your fragment behavior is not just the value, it's a key-value pair expected by `DiscoveryNodes` parsing and you require the format to be:
```json
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
```
Essentially, the optional start object IS read when parsing a standalone object but is NOT read when parsing the specific key/value format you include in `DiscoveryNodes` which omits the end object.

So ultimately my chief complaint here is that having the "field name" as part of this object is wrong; I should be able to have a standalone object that works without needing `DiscoveryNodes` wrapped around it.

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:605]: To be more clear/succinct, your representation of `DiscoveryNode` in JSON is this (optionally wrapped in curly braces) key value pair.
```json
{
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
}
```
or this because the object wrapper is optional, but requires the method caller to know that it's expecting a field name and object:
```json
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
```
This is not the expectation I've seen anywhere for XContentFragments which are usually standalone values (which may or may not be objects).

I believe it should either be:
{
  "name" : "name_1",
  "id" : "node_1",
  "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
  "transport_address" : "0.0.0.0:2",
  "attributes" : {
      "custom" : "PKU"
  }
}
```
or your original format with the object wrappers mandatory (even if you keep the fragment interface).

You're mixing-and-matching pairing of start object which relies heavily on implementation details of other classes.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:605]: @dbwiddis Thanks for being patient to write the detailed comment. Got your point on that we should be able to pass the inner nested object without the fieldname. But we need to create a parser which takes id as a parameter.
```
DiscoveryNode fromXContent(XContentParser parser, String nodeId)
```
I am assuming this will help address your concern.

Comment (shiv0408) [reply server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:612]: @dbwiddis Thanks for reviewing, I have added the the javadoc for this method.

File: libs/core/src/main/java/org/opensearch/core/common/transport/TransportAddress.java [hunk 6]
```diff
@@ -162,4 +164,22 @@ public String toString() {
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         return builder.value(toString());
     }
+
+    /**
+    * Converts a string in the format [hostname/ip]:[port] into a transport address.
+    * @throws UnknownHostException if the hostname or ip address is invalid
+    */
+    public static TransportAddress fromString(String address) throws UnknownHostException {
```
Comment (shwetathareja) [inline libs/core/src/main/java/org/opensearch/core/common/transport/TransportAddress.java:172]: You don't control the toString() you are using ```TransportAddress.toString()``` which calls ```NetworkAddress.format(address);``` . If toString() implementation changes, your fromString() will break.

**Comment by shwetathareja** [comment]
As discussed, it may not be scalable to keep on adding fromXContent to ephemeral objects which only have StreamInput/ StreamOutput implementation for node to node cluster state publication. Lets use that only serde. Also for other objects like ClusterState.Custom or Metadata.Custom where plugins can provides custom implementations, they may not have fromXContent method implemented and then it will break the remote state which is not desired.

**Comment by shiv0408** [comment]
Closing this as we are not gonna be using XContent for storing cluster state ephemeral objects brings overhead in maintenance of `writeTo`, `readFrom` and `toXContent` with additional `fromXContent`. We have added [RemoteWriteableEntity](https://github.com/opensearch-project/OpenSearch/blob/3d1fa9848f5a83e689c0eed1039613125d4f8fdf/server/src/main/java/org/opensearch/common/remote/RemoteWriteableEntity.java#L20)  with ser/de that can be done with InputStream as well.
So, discarding this PR and other ephemeral objects.
